File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        check = set()
        cur = head

        while cur:
            if cur in check:
                return True
            check.add(cur)
            cur = cur.next

        return False

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> old;
        old[NULL] = NULL;

        Node* cur = head;
        while (cur) {
            Node* copy = new Node(cur->val);
            old[cur] = copy;
            cur = cur->next;
        }

        cur = head;
        while (cur) {
            Node* copy = old[cur];
            copy->next = old[cur->next];
            copy->random = old[cur->random];
            cur = cur->next;
        }
        return old[head];
    }
};
